torch/distributed/nn/functional.py

```diff
@@ -219,19 +289,22 @@ def forward(ctx, group, tensor):
 
     @staticmethod
     def backward(ctx, *grad_outputs):
-        gxs = _AlltoAll.apply(ctx.group, *grad_outputs)
+        tensor_list = [
+            torch.empty_like(tensor) for tensor in grad_outputs
+        ]
+        gxs = _AlltoAll.apply(ctx.group, tensor_list, *grad_outputs)
         gx = torch.sum(torch.stack(gxs), dim=0)
         return (None, gx)
 
 
 class _AlltoAll(Function):
     @staticmethod
-    def forward(ctx, group, *tensors):
+    def forward(ctx, group, out_tensor_list, *tensors):
         ctx.group = group
-        out_tensor_list = [
-            torch.empty_like(tensors[i]) for i in range(dist.get_world_size(group=group))
+        ctx.input_tensor_size_list = [
+            tensors[i].size() for i in range(dist.get_world_size(group=group))
         ]
-        reqs = [None] * dist.get_world_size(group=group)
+        out_tensor_list = [tensor.contiguous() for tensor in out_tensor_list]
```

Comment:
I'm wondering if it is possible to avoid have `.contiguous` calls at all in the nn/functional.py APIs and deal with this in the sharded linear implementation itself?

For a general autograd enabled collective API it is better to throw an exception regarding contiguous tensors are not supported instead of adding overhead of the `contiguous` conversion.

Although, this might be tricky since the forward pass can be easily controlled in the sharded linear implementation but the backward pass is harder.

Reply:
Yeah, the test failure happens in the backward stage which might be out of our own control.